eureka2-core/src/main/java/com/netflix/eureka2/transport/base/BaseMessageConnection.java

```diff
@@ -267,29 +295,73 @@ public void call() {
                 .doOnCompleted(new Action0() {
                     @Override
                     public void call() {
-                        pendingAck.add(ack);
+                        pendingAckQueue.add(ack);
+                        metrics.incrementPendingAckCounter();
                         metrics.incrementOutgoingMessageCounter(message.getClass(), 1);
+
+                        // Connection might be closed when we serve this request, and since
+                        // pendingAckQueue and closed variables are not changed together atomically, we check
+                        // close status after adding item to pendingAckQueue, and optionally drain it.
+                        if (closed.get()) {
+                            drainPendingAckQueue();
```

Comment:
Yes, it is. I added a corresponding comment for drainPendingAckQueue()
method.

On Mon, May 4, 2015 at 3:19 PM, qiangdavidliu <notifications@github.com>
wrote:

> In
> eureka2-core/src/main/java/com/netflix/eureka2/transport/base/BaseMessageConnection.java
> <https://github.com/Netflix/eureka/pull/507#discussion_r29629664>:
>
> >                          metrics.incrementOutgoingMessageCounter(message.getClass(), 1);
> > +
> > +                        // Connection might be closed when we serve this request, and since
> > +                        // pendingAckQueue and closed variables are not changed together atomically, we check
> > +                        // close status after adding item to pendingAckQueue, and optionally drain it.
> > +                        if (closed.get()) {
> > +                            drainPendingAckQueue();
>
> Is this threadsafe? Seems like this can be called by multiple observable
> chains concurrently here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Netflix/eureka/pull/507/files#r29629664>.
>